transformer_engine/common/include/transformer_engine/fused_attention.h

```diff
@@ -0,0 +1,59 @@
+/*************************************************************************
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#ifndef TRANSFORMER_ENGINE_FUSED_ATTENTION_H_
+#define TRANSFORMER_ENGINE_FUSED_ATTENTION_H_
+
+#include "../common.h"
+#include "transformer_engine.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CUDNN_FRONTEND_UNUSED(X) ((void)X)
+
+enum class MHA_Layout { NOT_INTERLEAVED = 0, QKV_INTERLEAVED = 1, KV_INTERLEAVED = 2 };
+
+enum class MHA_Matrix {
+    Q_Matrix = 0,            // queries
+    K_Matrix = 1,            // keys
+    K_Matrix_Transpose = 2,  // keys transposed
+    V_Matrix = 3,            // values
+    V_Matrix_Transpose = 4,  // value matrix transposed
+    S_Matrix = 5,            // output of GEMM1
+    O_Matrix = 6,            // final output
+};
+
+enum class MHA_Bias_Type { NO_BIAS = 0, PRE_SCALE_BIAS = 1, POST_SCALE_BIAS = 2 };
+
+void nvte_cudnn_fused_attention_fwd(int64_t batch, int64_t max_q_seqlen, int64_t max_kv_seqlen,
```

Comment:
I really don't like this API - it forces the user to provide some random collection of tensors that are specific to a particular fused attention implementation (which it was supposed to abstract). The user should not need to know that a particular implementation needs `m` or `z_inv` or `softmax_aux` (whatever that is). If you need to provide a way to pass things between forward and backward then just pass a list of tensors and let the function populate it.

Going one by one:
 - `batch` - why do you need it? You already have ragged_offsets or actual_seqlens which should have a shape that includes batch size already
 - `max_q_seqlen`, `max_kv_seqlen` - that's probably OK to have
 - `total_seqs' - how is this different from the `batch` parameter?
 - `num_head` - why can't you get that from `qkv` tensor?
 - `head_size` - same as `num_head`
 - `scale_qk` - what is that?
 - `p_dropout` - OK
 - `is_causal_masking` - you already have some enums here, why not use an enum for the type to keep the API future proof?
 - `is_training` - OK, alternatively we could have a special inference API
 - `qkv_layout` - I would probably prefer us to extend NVTETensor to have strides
 - `bias_type` and `bias` - what is it?
 - `qkv` - OK
 - `m`, `z_inv`, `softmax_aux` - as mentioned at the beginning, I hate that. Please make something like `NVTETensor ** aux_outputs, size_t * num_aux_outputs` instead so that we can accommodate different implementations
 - `output` - OK
 - `q_ragged_offset`, `kv_ragged_offset`, `actual_q_seqlen`, `actual_kv_seqlen` - can't we combine those somehow, those seem redundant
 - `philox_unpack` - change the name to `rng_state` or something like that
 - `workspace`, `stream` - OK

Reply:
I've created a new API that's based on input/output tensor lists. There's still some tweaking to do, but here it is: https://github.com/cyanguwa/TransformerEngine/tree/fp8_flash_attn_newAPI